agent/consul/leader_connect_test.go

```diff
@@ -24,109 +25,117 @@ import (
 func TestLeader_SecondaryCA_Initialize(t *testing.T) {
 	t.Parallel()
 
-	masterToken := "8a85f086-dd95-4178-b128-e10902767c5c"
-
-	// Initialize primary as the primary DC
-	dir1, s1 := testServerWithConfig(t, func(c *Config) {
-		c.Datacenter = "primary"
-		c.ACLDatacenter = "primary"
-		c.Build = "1.6.0"
-		c.ACLsEnabled = true
-		c.ACLMasterToken = masterToken
-		c.ACLDefaultPolicy = "deny"
-	})
-	defer os.RemoveAll(dir1)
-	defer s1.Shutdown()
-
-	s1.tokens.UpdateAgentToken(masterToken, token.TokenSourceConfig)
-
-	testrpc.WaitForLeader(t, s1.RPC, "primary")
-
-	// secondary as a secondary DC
-	dir2, s2 := testServerWithConfig(t, func(c *Config) {
-		c.Datacenter = "secondary"
-		c.ACLDatacenter = "primary"
-		c.Build = "1.6.0"
-		c.ACLsEnabled = true
-		c.ACLDefaultPolicy = "deny"
-		c.ACLTokenReplication = true
-	})
-	defer os.RemoveAll(dir2)
-	defer s2.Shutdown()
-
-	s2.tokens.UpdateAgentToken(masterToken, token.TokenSourceConfig)
-	s2.tokens.UpdateReplicationToken(masterToken, token.TokenSourceConfig)
-
-	testrpc.WaitForLeader(t, s2.RPC, "secondary")
-
-	// Create the WAN link
-	joinWAN(t, s2, s1)
-
-	waitForNewACLs(t, s1)
-	waitForNewACLs(t, s2)
-
-	// Ensure s2 is authoritative.
-	waitForNewACLReplication(t, s2, structs.ACLReplicateTokens, 1, 1, 0)
-
-	// Wait until the providers are fully bootstrapped.
-	var (
-		caRoot            *structs.CARoot
-		secondaryProvider ca.Provider
-		intermediatePEM   string
-		err               error
-	)
-	retry.Run(t, func(r *retry.R) {
-		_, caRoot = s1.getCAProvider()
-		secondaryProvider, _ = s2.getCAProvider()
-		intermediatePEM, err = secondaryProvider.ActiveIntermediate()
-		require.NoError(r, err)
-
-		// Verify the root lists are equal in each DC's state store.
-		state1 := s1.fsm.State()
-		_, roots1, err := state1.CARoots(nil)
-		require.NoError(r, err)
-
-		state2 := s2.fsm.State()
-		_, roots2, err := state2.CARoots(nil)
-		require.NoError(r, err)
-		require.Len(r, roots1, 1)
-		require.Len(r, roots1, 1)
-		require.Equal(r, roots1[0].ID, roots2[0].ID)
-		require.Equal(r, roots1[0].RootCert, roots2[0].RootCert)
-		require.Empty(r, roots1[0].IntermediateCerts)
-		require.NotEmpty(r, roots2[0].IntermediateCerts)
-	})
-
-	// Have secondary sign a leaf cert and make sure the chain is correct.
-	spiffeService := &connect.SpiffeIDService{
-		Host:       "node1",
-		Namespace:  "default",
-		Datacenter: "primary",
-		Service:    "foo",
+	tests := []struct {
+		keyType string
+		keyBits int
+	}{
+		{connect.DefaultPrivateKeyType, connect.DefaultPrivateKeyBits},
+		{"rsa", 2048},
 	}
-	raw, _ := connect.TestCSR(t, spiffeService)
 
-	leafCsr, err := connect.ParseCSR(raw)
-	require.NoError(t, err)
-
-	leafPEM, err := secondaryProvider.Sign(leafCsr)
-	require.NoError(t, err)
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("%s-%d", tt.keyType, tt.keyBits), func(t *testing.T) {
```

Comment:
```suggestion
		tt := tt
		t.Run(fmt.Sprintf("%s-%d", tt.keyType, tt.keyBits), func(t *testing.T) {
```
I've gotten into the habit of doing this all the time so that the one time I make the subtest parallel I don't get unexpected test results where all of the subtests use the last testcase configuration.

Reply:
Does it only matter if you have `t.Parallel()` _inside_ the `t.Run`? Because tons of our suite have `t.Parallel()` at the top level.

I've added these for now although not super sure it's necessary if it only matters if you parallelise the inner test - that seems rare enough that we should maybe not optimize for it?

Reply:
It does only matter if the subtest itself is parallel (Parallel-inside-Run). I originally ran across this surprise in existing ACL tests that were masking a real bug since all but the last case were hidden.

I figured it didn't hurt to dupe the variable to avoid capturing the range variable proactively to avoid future surprises, but I am also not terribly attached to the practice if others think it's just overkill.

For reference the pattern `t.Run(.*\n.*t.Parallel(` occurs 229 times in our tests:
```
$ git grep -A 1 "t\.Run(" | grep -B 1 "t\.Parallel(" | grep -c Run
229
```

Reply:
Huh I had no idea we do that so often. Yeah I don't think there is harm in the copy as a practice. I wonder how many of those 229 are table tests or some other closure masking failures 🤔.

Reply:
My 2¢ is that its fine to always capture any table data when executing subtests. Better safe then thinking tests are working when there not.